pss/outbox/outbox.go

```diff
@@ -0,0 +1,130 @@
+package outbox
+
+import (
+	"errors"
+
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/pss/message"
+)
+
+type Config struct {
+	NumberSlots int
+	NumWorkers  int
+	Forward     forwardFunction
+}
+
+type Outbox struct {
+	forwardFunc forwardFunction
+	queue       []*outboxMsg
+	slots       chan int
+	process     chan int
+	workers     chan int
+	stopC       chan struct{}
+}
+
+type forwardFunction func(msg *message.Message) error
+
+var ErrOutboxFull = errors.New("outbox full")
+
+const defaultOutboxWorkers = 100
+
+func NewOutbox(config *Config) *Outbox {
+	if config.NumWorkers == 0 {
+		config.NumWorkers = defaultOutboxWorkers
+	}
+	outbox := &Outbox{
+		forwardFunc: config.Forward,
+		queue:       make([]*outboxMsg, config.NumberSlots),
+		slots:       make(chan int, config.NumberSlots),
+		process:     make(chan int),
+		workers:     make(chan int, config.NumWorkers),
+		stopC:       make(chan struct{}),
+	}
+	// fill up outbox slots
+	for i := 0; i < cap(outbox.slots); i++ {
+		outbox.slots <- i
+	}
+	return outbox
+}
+
+func (o *Outbox) Start() {
+	log.Info("Starting outbox")
+	go o.processOutbox()
+}
+
+func (o *Outbox) Stop() {
+	log.Info("Stopping outbox")
+	close(o.stopC)
+}
+
+// Enqueue a new element in the outbox if there is any slot available.
+// Then send it to process. This method is blocking in the process channel!
+func (o *Outbox) Enqueue(outboxMsg *outboxMsg) error {
+	// first we try to obtain a slot in the outbox
+	select {
+	case slot := <-o.slots:
+		o.queue[slot] = outboxMsg
+		metrics.GetOrRegisterGauge("pss.outbox.len", nil).Update(int64(o.len()))
+		// we send this message slot to process
+		select {
+		case <-o.stopC:
+		case o.process <- slot:
```

Comment:
Why is process channel needed? It is read in a separate goroutine just to start a new goroutine. I believe that the same goroutine can be created here, without the intermediate process channel as it is not buffered. In that case workers channel needs to stay as a field in Outbox struct and the whole processOutbox function can be removed.

Reply:
I think I can extract the code inside processOutbox to some processSlot, removing the process channel. But I'm not sure how to do the requeue when the forwarding fails. Before we sent a message to process, so the main routine in processOutbox would eventually take the signal and spawn a new thread. Now, it doesn't make sense that the worker spawn itself a new worker. It should free the worker slot so a new message is processed and put its message in a queue to be retried eventually. So we need an intermediate channel to requeue the failed messages. 
As the workers can receive jobs from two places (newly arrived messages and retried messages) I think we need a master routine that controls the spawning of threads.
Anyway, that was the design without limit in workers, not sure if by adding this limit the design should change, so it is out of the scope of this PR.

Reply:
I see, I missed the requeue send to the process channel. Thanks.